integration/service/create_test.go

```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+
+	// verify that the task has the swap option set in the task object
+	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
+		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
+	}
+
+	// verify that the service also has the swap set in the spec.
+	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+	assert.NilError(t, err)
+	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
+		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
+	}
+}
+
+func TestCreateServiceWithSwapBytesAndNoMemoryLimitFails(t *testing.T) {
+	skip.If(
+		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
+		"setting service swappiness is unsupported before api v1.40",
+	)
+
+	defer setupTest(t)()
+	d := swarm.NewSwarm(t, testEnv)
+	defer d.Stop(t)
+	client := d.NewClientT(t)
+	defer client.Close()
+	swap := int64(10 * units.MiB)
+
+	serviceOpts := func(spec *swarmtypes.ServiceSpec) {
+		if spec.TaskTemplate.Resources == nil {
+			spec.TaskTemplate.Resources = &swarmtypes.ResourceRequirements{}
+		}
+		spec.TaskTemplate.Resources.SwapBytes = &swap
+	}
+
+	spec := swarm.CreateServiceSpec(t, serviceOpts)
+	_, err := client.ServiceCreate(context.Background(), spec, types.ServiceCreateOptions{})
+
+	assert.ErrorContains(t, err, "Please set the memory limit when setting the swap size, see usage")
+}
+
+func TestCreateServiceWithMemorySwappiness(t *testing.T) {
+	skip.If(
+		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
+		"setting service swappiness is unsupported before api v1.40",
+	)
+
+	defer setupTest(t)()
+	d := swarm.NewSwarm(t, testEnv)
+	defer d.Stop(t)
+	client := d.NewClientT(t)
+	defer client.Close()
+
+	ctx := context.Background()
+	swappiness := int64(88)
+
+	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+		swarm.ServiceWithMemorySwappiness(swappiness))
+
+	// verify that the container has the swappiness option set
+	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+	assert.NilError(t, err)
+	if assert.Check(t, ctnr.HostConfig.MemorySwappiness != nil) {
+		assert.Equal(t, *ctnr.HostConfig.MemorySwappiness, swappiness)
+	}
+
+	// verify that the task has the swappiness option set in the task object
+	if assert.Check(t, task.Spec.Resources.MemorySwappiness != nil) {
+		assert.Equal(t, *task.Spec.Resources.MemorySwappiness, swappiness)
+	}
+
+	// verify that the service also has the swappiness set in the spec.
+	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+	assert.NilError(t, err)
+	if assert.Check(t, service.Spec.TaskTemplate.Resources.MemorySwappiness != nil) {
+		assert.Equal(t, *service.Spec.TaskTemplate.Resources.MemorySwappiness, swappiness)
+	}
+}
+
+// Tests that if the swappiness is not set, it should be reported as nil
+func TestCreateServiceWithNoMemorySwappiness(t *testing.T) {
```

Comment:
Could be something like this;

I noticed that we may need some testing for out-of-range values (`swappiness  -1`, `swappiness 200`)


```go
func TestCreateServiceMemorySwappiness(t *testing.T) {
	skip.If(
		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
		"setting service swap is unsupported before api v1.40",
	)
	skip.If(t, !testEnv.DaemonInfo.SwapLimit)

	defer setupTest(t)()
	d := swarm.NewSwarm(t, testEnv)
	defer d.Stop(t)
	client := d.NewClientT(t)
	defer client.Close()
	ctx := context.Background()

	toPtr := func(v int64) *int64 { return &v }

	tests := []struct {
		doc        string
		swappiness *int64
		expected   int64
		expectedCg string
	}{
		{doc: "default", expectedCg: "60"},
		{doc: "zero memory-swappiness", swappiness: toPtr(0), expectedCg: "0"},
		{doc: "memory-swappiness", swappiness: toPtr(88), expectedCg: "88"},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.doc, func(t *testing.T) {

			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwappiness(tc.swappiness))
			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
			assert.NilError(t, err)

			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
			assert.NilError(t, err)

			if tc.swappiness == nil {
				assert.Check(t, is.Nil(ctnr.HostConfig.Resources.MemorySwappiness))
				assert.Check(t, is.Nil(task.Spec.Resources.MemorySwappiness))
				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.MemorySwappiness))
			} else {
				assert.Check(t, is.Equal(*ctnr.HostConfig.Resources.MemorySwappiness, *tc.swappiness))
				assert.Check(t, is.Equal(*task.Spec.Resources.MemorySwappiness, *tc.swappiness))
				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.MemorySwappiness, *tc.swappiness))
			}

			res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.swappiness"})
			assert.NilError(t, err)
			assert.Assert(t, is.Len(res.Stderr(), 0))
			assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)

		})
	}
}
```

Reply:
Full diff

<details>

```patch
diff --git a/integration/internal/swarm/service.go b/integration/internal/swarm/service.go
index 365f345b90..8d95e696b8 100644
--- a/integration/internal/swarm/service.go
+++ b/integration/internal/swarm/service.go
@@ -176,17 +176,17 @@ func ServiceWithSysctls(sysctls map[string]string) ServiceSpecOpt {
 
 // ServiceWithMemorySwap sets the memory swap value of the service's TaskSpec
 // it also requires setting a memory limit, since the engine enforces that
-func ServiceWithMemorySwap(swap, memoryLimit int64) ServiceSpecOpt {
+func ServiceWithMemorySwap(swap *int64, memoryLimit int64) ServiceSpecOpt {
 	return withEnsureResourceLimits(func(spec *swarmtypes.ServiceSpec) {
-		spec.TaskTemplate.Resources.SwapBytes = &swap
+		spec.TaskTemplate.Resources.SwapBytes = swap
 		spec.TaskTemplate.Resources.Limits.MemoryBytes = memoryLimit
 	})
 }
 
 // ServiceWithMemorySwappiness sets the memory swappiness value of the service's ContainerSpec
-func ServiceWithMemorySwappiness(swappiness int64) ServiceSpecOpt {
+func ServiceWithMemorySwappiness(swappiness *int64) ServiceSpecOpt {
 	return withEnsureResourceLimits(func(spec *swarmtypes.ServiceSpec) {
-		spec.TaskTemplate.Resources.MemorySwappiness = &swappiness
+		spec.TaskTemplate.Resources.MemorySwappiness = swappiness
 	})
 }
 
diff --git a/integration/service/create_test.go b/integration/service/create_test.go
index 8b51465f5f..fd98444c53 100644
--- a/integration/service/create_test.go
+++ b/integration/service/create_test.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"strings"
 	"testing"
 	"time"
 
@@ -13,11 +14,11 @@ import (
 	swarmtypes "github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/api/types/versions"
 	"github.com/docker/docker/client"
+	"github.com/docker/docker/integration/internal/container"
 	"github.com/docker/docker/integration/internal/network"
 	"github.com/docker/docker/integration/internal/swarm"
 	"github.com/docker/docker/internal/test/daemon"
 	"github.com/docker/docker/internal/test/request"
-	"github.com/docker/docker/pkg/sysinfo"
 	"github.com/docker/go-units"
 	"gotest.tools/assert"
 	is "gotest.tools/assert/cmp"
@@ -192,7 +193,7 @@ func TestCreateWithDuplicateNetworkNames(t *testing.T) {
 
 	// Create Service with the same name
 	serviceName := "top_" + t.Name()
-	serviceID, _ := createServiceWithSingleInstanceAndWaitConverge(context.Background(), t, d, client,
+	serviceID, _ := createServiceAndConverge(context.Background(), t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithNetwork(name),
 	)
@@ -241,7 +242,7 @@ func TestCreateServiceSecretFileMode(t *testing.T) {
 	assert.NilError(t, err)
 
 	serviceName := "TestService_" + t.Name()
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+	serviceID, task := createServiceAndConverge(ctx, t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithCommand([]string{"/bin/sh", "-c", "ls -l /etc/secret || /bin/top"}),
 		swarm.ServiceWithSecret(&swarmtypes.SecretReference{
@@ -293,7 +294,7 @@ func TestCreateServiceConfigFileMode(t *testing.T) {
 	assert.NilError(t, err)
 
 	serviceName := "TestService_" + t.Name()
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+	serviceID, task := createServiceAndConverge(ctx, t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithCommand([]string{"/bin/sh", "-c", "ls -l /etc/config || /bin/top"}),
 		swarm.ServiceWithConfig(&swarmtypes.ConfigReference{
@@ -368,7 +369,7 @@ func TestCreateServiceSysctls(t *testing.T) {
 		// store the map we're going to be using everywhere.
 		expectedSysctls := map[string]string{"net.ipv4.ip_nonlocal_bind": expected}
 
-		serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+		serviceID, task := createServiceAndConverge(ctx, t, d, client,
 			swarm.ServiceWithSysctls(expectedSysctls))
 
 		// we're going to check 3 things:
@@ -413,232 +414,174 @@ func TestCreateServiceSysctls(t *testing.T) {
 	}
 }
 
-func TestCreateServiceWithMemorySwap(t *testing.T) {
+func TestCreateServiceMemorySwap(t *testing.T) {
 	skip.If(
 		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
 		"setting service swap is unsupported before api v1.40",
 	)
+	skip.If(t, !testEnv.DaemonInfo.SwapLimit)
 
 	defer setupTest(t)()
 	d := swarm.NewSwarm(t, testEnv)
 	defer d.Stop(t)
 	client := d.NewClientT(t)
 	defer client.Close()
-
 	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(1 * units.MiB)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-	// verify that the container has the swap option set - provided that the host
-	// supports it! see
-	// https://github.com/moby/moby/blob/v17.03.2-ce/daemon/daemon_unix.go#L290-L294
-	if sysinfo.New(true).SwapLimit {
-		ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-		assert.NilError(t, err)
-		assert.Equal(t, ctnr.HostConfig.MemorySwap, memoryLimit+swap)
-	}
-
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
-
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
-	}
-}
-
-func TestCreateServiceWithMemorySwapEqualToZero(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swap is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
 
-	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(0)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-	// verify that the container has the swap option set - provided that the host
-	// supports it, same as `TestCreateServiceWithMemorySwap` above
-	if sysinfo.New(true).SwapLimit {
-		ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-		assert.NilError(t, err)
-		assert.Equal(t, ctnr.HostConfig.MemorySwap, memoryLimit)
+	toPtr := func(v int64) *int64 { return &v }
+
+	tests := []struct {
+		doc         string
+		memorySwap  *int64
+		memoryLimit int64
+		expected    int64
+		expectedCg  string
+		expectedErr string
+	}{
+		{
+			doc: "default",
+		},
+		{
+			doc:         "memory-limit and memory-swap",
+			memorySwap:  toPtr(1 * units.MiB),
+			memoryLimit: 20 * units.MiB,
+			expected:    21 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 21*units.MiB),
+		},
+		{
+			doc:         "memory-limit",
+			memoryLimit: 20 * units.MiB,
+			expected:    40 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 40*units.MiB),
+		},
+		{
+			doc:         "memory-limit and zero memory-swap",
+			memorySwap:  toPtr(0),
+			memoryLimit: 20 * units.MiB,
+			expected:    20 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 20*units.MiB),
+		},
+		{
+			doc:         "memory-limit and unlimited memory-swap",
+			memorySwap:  toPtr(-1),
+			memoryLimit: 20 * units.MiB,
+			expected:    -1,
+			expectedCg:  "", // maximum value depends on pagesize and host configuration
+		},
+		{
+			doc:         "memory-swap only",
+			memorySwap:  toPtr(10 * units.MiB),
+			expectedErr: "Please set the memory limit when setting the swap size, see usage",
+		},
 	}
 
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
+			if tc.expectedErr != "" {
+				_, err := client.ServiceCreate(
+					context.Background(),
+					swarm.CreateServiceSpec(t, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit)),
+					types.ServiceCreateOptions{},
+				)
+				assert.ErrorContains(t, err, tc.expectedErr)
+				return
+			}
+
+			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit))
+			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+			assert.NilError(t, err)
+
+			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+			assert.NilError(t, err)
+
+			if tc.memorySwap == nil {
+				assert.Check(t, is.Nil(task.Spec.Resources.SwapBytes))
+				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.SwapBytes))
+			} else {
+				assert.Check(t, is.Equal(*task.Spec.Resources.SwapBytes, *tc.memorySwap))
+				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.SwapBytes, *tc.memorySwap))
+			}
+
+			assert.Check(t, is.Equal(ctnr.HostConfig.Resources.MemorySwap, tc.expected))
+
+			if tc.expectedCg != "" {
+				res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes"})
+				assert.NilError(t, err)
+				assert.Assert(t, is.Len(res.Stderr(), 0))
+				assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)
+			}
 
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
+		})
 	}
 }
 
-func TestCreateServiceWithUnlimitedMemorySwap(t *testing.T) {
+func TestCreateServiceMemorySwappiness(t *testing.T) {
 	skip.If(
 		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
 		"setting service swap is unsupported before api v1.40",
 	)
+	skip.If(t, !testEnv.DaemonInfo.SwapLimit)
 
 	defer setupTest(t)()
 	d := swarm.NewSwarm(t, testEnv)
 	defer d.Stop(t)
 	client := d.NewClientT(t)
 	defer client.Close()
-
-	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(-1)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-
-	// verify that the container has the swap option set to -1
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	assert.Equal(t, ctnr.HostConfig.MemorySwap, swap)
-
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
-
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
-	}
-}
-
-func TestCreateServiceWithSwapBytesAndNoMemoryLimitFails(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-	swap := int64(10 * units.MiB)
-
-	serviceOpts := func(spec *swarmtypes.ServiceSpec) {
-		if spec.TaskTemplate.Resources == nil {
-			spec.TaskTemplate.Resources = &swarmtypes.ResourceRequirements{}
-		}
-		spec.TaskTemplate.Resources.SwapBytes = &swap
-	}
-
-	spec := swarm.CreateServiceSpec(t, serviceOpts)
-	_, err := client.ServiceCreate(context.Background(), spec, types.ServiceCreateOptions{})
-
-	assert.ErrorContains(t, err, "Please set the memory limit when setting the swap size, see usage")
-}
-
-func TestCreateServiceWithMemorySwappiness(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-
 	ctx := context.Background()
-	swappiness := int64(88)
 
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwappiness(swappiness))
-
-	// verify that the container has the swappiness option set
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	if assert.Check(t, ctnr.HostConfig.MemorySwappiness != nil) {
-		assert.Equal(t, *ctnr.HostConfig.MemorySwappiness, swappiness)
+	toPtr := func(v int64) *int64 { return &v }
+
+	tests := []struct {
+		doc        string
+		swappiness *int64
+		expected   int64
+		expectedCg string
+	}{
+		{doc: "default", expectedCg: "60"},
+		{doc: "zero memory-swappiness", swappiness: toPtr(0), expectedCg: "0"},
+		{doc: "memory-swappiness", swappiness: toPtr(88), expectedCg: "88"},
 	}
 
-	// verify that the task has the swappiness option set in the task object
-	if assert.Check(t, task.Spec.Resources.MemorySwappiness != nil) {
-		assert.Equal(t, *task.Spec.Resources.MemorySwappiness, swappiness)
-	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
 
-	// verify that the service also has the swappiness set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.MemorySwappiness != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.MemorySwappiness, swappiness)
-	}
-}
+			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwappiness(tc.swappiness))
+			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+			assert.NilError(t, err)
 
-// Tests that if the swappiness is not set, it should be reported as nil
-func TestCreateServiceWithNoMemorySwappiness(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
+			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+			assert.NilError(t, err)
 
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-
-	ctx := context.Background()
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client)
+			if tc.swappiness == nil {
+				assert.Check(t, is.Nil(ctnr.HostConfig.Resources.MemorySwappiness))
+				assert.Check(t, is.Nil(task.Spec.Resources.MemorySwappiness))
+				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.MemorySwappiness))
+			} else {
+				assert.Check(t, is.Equal(*ctnr.HostConfig.Resources.MemorySwappiness, *tc.swappiness))
+				assert.Check(t, is.Equal(*task.Spec.Resources.MemorySwappiness, *tc.swappiness))
+				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.MemorySwappiness, *tc.swappiness))
+			}
 
-	// shouldn't be set on the container
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	assert.Check(t, ctnr.HostConfig.MemorySwappiness == nil)
+			res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.swappiness"})
+			assert.NilError(t, err)
+			assert.Assert(t, is.Len(res.Stderr(), 0))
+			assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)
 
-	// nor on the task
-	if task.Spec.Resources != nil {
-		assert.Check(t, task.Spec.Resources.MemorySwappiness == nil)
-	}
-
-	// nor on the service's spec
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if service.Spec.TaskTemplate.Resources != nil {
-		assert.Check(t, service.Spec.TaskTemplate.Resources.MemorySwappiness == nil)
+		})
 	}
 }
 
-func createServiceWithSingleInstanceAndWaitConverge(ctx context.Context, t *testing.T, d *daemon.Daemon, client *client.Client, opts ...swarm.ServiceSpecOpt) (serviceID string, task swarmtypes.Task) {
+func createServiceAndConverge(ctx context.Context, t *testing.T, d *daemon.Daemon, client *client.Client, opts ...swarm.ServiceSpecOpt) (serviceID string, task swarmtypes.Task) {
 	serviceID = swarm.CreateService(t, d, opts...)
 
 	// wait for the service to converge to 1 running task as expected
 	poll.WaitOn(t, swarm.RunningTasksCount(client, serviceID, 1))
 
 	// retrieve the task
-	filter := filters.NewArgs()
-	filter.Add("service", serviceID)
 	tasks, err := client.TaskList(ctx, types.TaskListOptions{
-		Filters: filter,
+		Filters: filters.NewArgs(filters.KeyValuePair{Key: "service", Value: serviceID}),
 	})
 	assert.NilError(t, err)
 	assert.Check(t, is.Equal(len(tasks), 1))
```

</details>

Reply:
👍